Add toolchain option to cross-build

Current cross-build uses the toolchain installed in the host so, we need to add user custom toolchain option to cross-build.

hi, if  target-compiler-triple=aarch64-none-linux-gnu will broken compiler that  output error
```
    /usr/bin/ld: unrecognised emulation mode: aarch64linux
    Supported emulations: elf_x86_64 elf32_x86_64 elf_i386 elf_iamcu elf_l1om elf_k1om i386pep i386pe
    clang: error: linker command failed with exit code 1 (use -v to see invocation)
    gmake[1]: *** [CMakeFiles/cmTC_0ca6b.dir/build.make:100: cmTC_0ca6b] Error 1
    gmake[1]: Leaving directory '/home/seanchann/source/opensrc/flutter-extension/decodetest/build/elinux/arm64/release/CMakeFiles/CMakeTmp'
    gmake: *** [Makefile:127: cmTC_0ca6b/fast] Error 2
```

--- a/lib/commands/build.dart
+++ b/lib/commands/build.dart
@@ -61,6 +61,11 @@ class BuildPackageCommand extends BuildSubCommand
           'if the current host and target architectures are different.',
     );
     argParser.addOption(
+      'target-toolchain',
+      defaultsTo: '/',
+      help: 'The toolchain path for Clang.',
+    );
+    argParser.addOption(
       'system-include-directories',
       defaultsTo: null,
       help:
@@ -120,8 +125,9 @@ class BuildPackageCommand extends BuildSubCommand
       targetBackendType: stringArg('target-backend-type'),
       targetCompilerTriple: stringArg('target-compiler-triple'),
       targetSysroot: stringArg('target-sysroot'),
-      systemIncludeDirectories: stringArg('system-include-directories'),
       targetCompilerFlags: stringArg('target-compiler-flags'),
+      targetToolchain: stringArg('target-toolchain'),
+      systemIncludeDirectories: stringArg('system-include-directories'),
     );
     validateBuild(eLinuxBuildInfo);
     displayNullSafetyMode(buildInfo);

--- a/lib/elinux_build_target.dart
+++ b/lib/elinux_build_target.dart
@@ -357,6 +357,7 @@ class NativeBundle {
     final String targetCompilerTriple = buildInfo.targetCompilerTriple;
     final String targetSysroot = buildInfo.targetSysroot;
     final String targetCompilerFlags = buildInfo.targetCompilerFlags;
+    final String targetToolchain = buildInfo.targetToolchain;
     final String systemIncludeDirectories = buildInfo.systemIncludeDirectories;
     RunResult result = await _processUtils.run(
       <String>[
@@ -378,7 +379,12 @@ class NativeBundle {
         eLinuxDir.path,
       ],
       workingDirectory: outputDir.path,
-      environment: <String, String>{'CC': 'clang', 'CXX': 'clang++'},
+      environment: (targetToolchain == null)
+          ? <String, String>{'CC': 'clang', 'CXX': 'clang++'}
+          : <String, String>{
+              'CC': '$targetToolchain/bin/clang',
+              'CXX': '$targetToolchain/bin/clang++'
+            },
     );
     if (result.exitCode != 0) {
       throwToolExit('Failed to cmake:\n$result');

--- a/lib/elinux_builder.dart
+++ b/lib/elinux_builder.dart
@@ -37,8 +37,9 @@ class ELinuxBuildInfo {
     @required this.targetBackendType,
     @required this.targetCompilerTriple,
     @required this.targetSysroot,
-    @required this.systemIncludeDirectories,
     @required this.targetCompilerFlags,
+    @required this.targetToolchain,
+    @required this.systemIncludeDirectories,
   })  : assert(targetArch != null),
         assert(targetBackendType != null);
 
@@ -47,8 +48,9 @@ class ELinuxBuildInfo {
   final String targetBackendType;
   final String targetCompilerTriple;
   final String targetSysroot;
-  final String systemIncludeDirectories;
   final String targetCompilerFlags;
+  final String targetToolchain;
+  final String systemIncludeDirectories;
 }
 
 /// See:

--- a/lib/elinux_device.dart
+++ b/lib/elinux_device.dart
@@ -299,8 +299,9 @@ class ELinuxDevice extends Device {
       targetBackendType: _backendType,
       targetCompilerTriple: null,
       targetSysroot: '/',
-      systemIncludeDirectories: null,
       targetCompilerFlags: null,
+      targetToolchain: null,
+      systemIncludeDirectories: null,
     );
     await ELinuxBuilder.buildBundle(
       project: project,

